src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -149,34 +149,34 @@ async def mypy_typecheck(
     #     `python_version`. We don't want to make the user set this up. (If they do, MyPy will use
     #     `python_version`, rather than defaulting to the executing interpreter).
     #
-    #     We only apply these optimizations if the user did not configure
-    #     `--mypy-interpreter-constraints`, and if we know that there are no Py35 or Py27
-    #     constraints. If they use Py27 or Py35, this implies that they're not using Py36+ syntax,
-    #     so it's fine to use the Py35 parser. We want the loosest constraints possible to make it
-    #     more flexible to install MyPy.
     #  * We must resolve third-party dependencies. This should use whatever the actual code's
-    #     constraints are. The constraints for the tool can be different than for the requirements.
+    #     constraints are. However, PEX will error if they are not compatible
+    #     with the interpreter being used to run MyPy. So, we should generally align the tool PEX
+    #     constraints with the requirements constraints.
+    #
+    #     However, it's possible for the requirements constraints to be different, e.g. if there is
+    #     Python 2 code. If any of the requirements only have Python 2 wheels and they are not
+    #     compatible with Python 3, then Pex will error about missing wheels. So, in this case, we
+    #     set `PEX_IGNORE_ERRORS`, which will avoid erroring, but may result in MyPy complaining
+    #     that it cannot find certain distributions.
+    #
     #  * The runner Pex should use the same constraints as the tool Pex.
-    all_interpreter_constraints = PexInterpreterConstraints.create_from_compatibility_fields(
+    code_interpreter_constraints = PexInterpreterConstraints.create_from_compatibility_fields(
         (
             tgt[PythonInterpreterCompatibility]
-            for tgt in itertools.chain(
-                typechecked_transitive_targets.closure, plugin_transitive_targets.closure
-            )
+            for tgt in typechecked_transitive_targets.closure
             if tgt.has_field(PythonInterpreterCompatibility)
         ),
         python_setup,
     )
-    if not mypy.options.is_default("interpreter_constraints"):
-        tool_interpreter_constraints = mypy.interpreter_constraints
-    elif all_interpreter_constraints.requires_python38_or_newer():
-        tool_interpreter_constraints = ("CPython>=3.8",)
-    elif all_interpreter_constraints.requires_python37_or_newer():
-        tool_interpreter_constraints = ("CPython>=3.7",)
-    elif all_interpreter_constraints.requires_python36_or_newer():
-        tool_interpreter_constraints = ("CPython>=3.6",)
-    else:
-        tool_interpreter_constraints = mypy.interpreter_constraints
+    tool_interpreter_constraints = (
+        code_interpreter_constraints
+        if (
+            mypy.options.is_default("interpreter_constraints")
+            and code_interpreter_constraints.includes_python2() is False
```

Comment:
This conditional is a little confusing to me. Could we leave a comment describing the result? Or is that already covered above?

Reply:
I can factor it out as a distinct variable. Good suggestion.